File: python_bindings_jupedsim/anticipation_velocity_model.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
#include "AnticipationVelocityModel.hpp"
#include "OperationalModel.hpp"
#include "type_casters.hpp" // IWYU pragma: keep

#include <pybind11/cast.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h> // IWYU pragma: keep

#include <cstdint>

namespace py = pybind11;

void init_anticipation_velocity_model(py::module_& m)
{
    py::class_<AnticipationVelocityModel, OperationalModel, py::smart_holder>(
        m, "AnticipationVelocityModel")
        .def(
            py::init<double, uint64_t>(),
            py::kw_only(),
            py::arg("pushout_strength") = 0.3,
            py::arg("rng_seed"));
    const AnticipationVelocityModel::State d{};
    py::class_<AnticipationVelocityModel::State>(m, "AnticipationVelocityModelState")
        .def(
            py::init([](Point position,
                        Point orientation,
                        double strengthNeighborRepulsion,
                        double rangeNeighborRepulsion,
                        double wallBufferDistance,
                        double anticipationTime,
                        double reactionTime,
                        Point velocity,
                        double timeGap,
                        double desiredSpeed,
                        double radius) {
                return AnticipationVelocityModel::State{
                    .position = position,
                    .orientation = orientation,
                    .strengthNeighborRepulsion = strengthNeighborRepulsion,
                    .rangeNeighborRepulsion = rangeNeighborRepulsion,
                    .wallBufferDistance = wallBufferDistance,
                    .anticipationTime = anticipationTime,
                    .reactionTime = reactionTime,
                    .velocity = velocity,
                    .timeGap = timeGap,
                    .v0 = desiredSpeed,
                    .radius = radius};
            }),
            py::kw_only(),
            py::arg("position") = d.position,
            py::arg("orientation") = d.orientation,
            py::arg("strength_neighbor_repulsion") = d.strengthNeighborRepulsion,
            py::arg("range_neighbor_repulsion") = d.rangeNeighborRepulsion,
            py::arg("wall_buffer_distance") = d.wallBufferDistance,
            py::arg("anticipation_time") = d.anticipationTime,
            py::arg("reaction_time") = d.reactionTime,
            py::arg("velocity") = d.velocity,
            py::arg("time_gap") = d.timeGap,
            py::arg("desired_speed") = d.v0,
            py::arg("radius") = d.radius)
        .def_readwrite("position", &AnticipationVelocityModel::State::position)
        .def_readwrite("orientation", &AnticipationVelocityModel::State::orientation)
        .def_readwrite(
            "strength_neighbor_repulsion",
            &AnticipationVelocityModel::State::strengthNeighborRepulsion)
        .def_readwrite(
            "range_neighbor_repulsion", &AnticipationVelocityModel::State::rangeNeighborRepulsion)
        .def_readwrite(
            "wall_buffer_distance", &AnticipationVelocityModel::State::wallBufferDistance)
        .def_readwrite("anticipation_time", &AnticipationVelocityModel::State::anticipationTime)
        .def_readwrite("reaction_time", &AnticipationVelocityModel::State::reactionTime)
        .def_readwrite("velocity", &AnticipationVelocityModel::State::velocity)
        .def_readwrite("time_gap", &AnticipationVelocityModel::State::timeGap)
        .def_readwrite("desired_speed", &AnticipationVelocityModel::State::v0)
        .def_readwrite("radius", &AnticipationVelocityModel::State::radius);
}

File: python_modules/jupedsim/jupedsim/models/anticipation_velocity_model.py
# SPDX-License-Identifier: LGPL-3.0-or-later
"""Anticipation Velocity Model (AVM).

The AVM incorporates pedestrian anticipation, divided into three phases:
perception of the current situation, prediction of future situations and
strategy selection leading to action. A general description of the AVM can be
found in the originating publication
https://doi.org/10.1016/j.trc.2021.103464

The model-level parameters (``pushout_strength`` and ``rng_seed``) are carried
by the model instance, which is passed to the simulation:

.. code:: python

    sim = jupedsim.Simulation(
        model=jupedsim.AnticipationVelocityModel(rng_seed=1234),
        geometry=...,
    )
    sim.add_agent(
        journey_id=journey_id,
        stage_id=stage_id,
        state=jupedsim.AnticipationVelocityModelState(position=(1.0, 1.0)),
    )

.. warning::

    The model instance is consumed by the ``Simulation`` constructor and must
    not be reused afterwards.

:class:`AnticipationVelocityModel` exposes the model-level parameters as
keyword-only constructor arguments with sensible defaults:
``pushout_strength`` and ``rng_seed``.

:class:`AnticipationVelocityModelState` exposes the complete per-agent state
of the model as keyword-only constructor arguments with sensible defaults:
``position``, ``orientation``, ``strength_neighbor_repulsion``,
``range_neighbor_repulsion``, ``wall_buffer_distance``,
``anticipation_time``, ``reaction_time``, ``velocity``, ``time_gap``,
``desired_speed`` and ``radius``.
"""

import jupedsim.native as py_jps

AnticipationVelocityModel = py_jps.AnticipationVelocityModel
AnticipationVelocityModelState = py_jps.AnticipationVelocityModelState

__all__ = [
    "AnticipationVelocityModel",
    "AnticipationVelocityModelState",
]
